Native Language Server integration with PM

hubertp opened the pull request:
### Pull Request Description

The first attempt at
1. Building `enso` native-image that includes full Language Server
2. Integrating the executable as an experimental feature during project startup

The change (for now) assumes that `enso` executable appears in Enso's default `engines` directory.
To build and run the new integration one has to
1. `engine-runner/buildNativeImage`
2. Run PM

The Runner will automatically detect the presence of a native image and attempt to run LS in AOT mode.

This change also adds a copy of some of logback's code (a simple socket server) as it was impossible to debug serialization bugs without some additional logging. Potentially to be removed in the final PR.

This is by no means a final change in this area. We need to make it possible to include all other `Standard` libs, at least. But it's a step forward that allows devs to experiment.

### Important Notes

There is a change in `JsonConnectionController` which only sets the controller as initialized when all resources are set up. Previously, it seems, the asynchronous setup could lead to some race conditions.

Exchange between LS and GUI (before):
![Screenshot from 2024-12-30 15-17-30](https://github.com/user-attachments/assets/38ed484b-aec6-4694-8c59-76c9f69f776b)

(after)
![Screenshot from 2024-12-30 15-20-08](https://github.com/user-attachments/assets/c0f0d343-ad38-4a89-8658-a2ba51699d58)

As one can see, speedup in startup is rather massive.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

hubertp committed:
Native Language Server integration with PM

The first attempt at
a) building `enso` native-image that includes full Language Server
b) integrating the executable as an experimental feature during project
startup

The change (for now) assumes that `enso` executable appears in Enso's
default `engines` directory.
To build and run the new integration one has to
a) `engine-runner/buildNativeImage`
b) run PM with `--native-language-server`

This change also adds a copy of some of logback's code (`SocketAppender`
or a simple socket server`) as it was impossible to debug serialization
bugs without some additional logging.

hubertp committed:
nit

hubertp committed:
Add missing reflect config

Otherwise Akka and other code is simply dead-code eliminated in
native-image.

hubertp committed:
Ensure proper Context construction

When message transport is missing, context should be properly built,
even in AOT mode.

hubertp committed:
Add missing Akka entries in reflect-config

hubertp committed:
nit

hubertp committed:
Merge branch 'develop' into wip/hubert/11721-language-server-native

hubertp committed:
Partial revert

hubertp committed:
Fix service registration

Problems revealed after a clean build

hubertp committed:
Merge branch 'develop' into wip/hubert/11721-language-server-native

hubertp committed:
Fix Base_Tests in native-image

hubertp committed:
Fix compilation issues

hubertp committed:
Drop Truffle dependency

hubertp committed:
fmt

hubertp committed:
Merge branch 'develop' into wip/hubert/11721-language-server-native

JaroslavTulach commented on build.sbt:
```diff
@@ -2170,11 +2170,14 @@ lazy val `engine-common` = project
         "org.graalvm.polyglot" % "polyglot"  % graalMavenPackagesVersion,
         "org.slf4j"            % "slf4j-api" % slf4jVersion
-      )
+      ) ++ GraalVM.modules.map(_.withConfigurations(Some(Runtime.name)))
```
The ScalaDoc of `engine-common` project which is available few lines above this change notes:

> As such it needs to depend on `org.graalvm.polyglot` package. Otherwise
> its dependencies shall be limited

As far as I can see [GraalVM.modules](https://github.com/enso-org/enso/blob/develop/project/GraalVM.scala#L29) contains Truffle API and that's not needed to boot `Context` - in fact that is an API that users (those not implementing a language) of the GraalVM shouldn't see at all.

What is the reason for these dependency changes?

JaroslavTulach commented on build.sbt:
```diff
@@ -2280,16 +2283,21 @@ lazy val `language-server` = (project in file("engine/language-server"))
+        "commons-io"             % "commons-io"              % commonsIoVersion,
+        "com.google.flatbuffers" % "flatbuffers-java"        % flatbuffersVersion,
+        "org.eclipse.jgit"       % "org.eclipse.jgit"        % jgitVersion,
+        "org.netbeans.api"       % "org-openide-util-lookup" % netbeansApiVersion
```
I don't think we need `org-openide-util-lookup` in the runtime. There is:
- #10972
- however the `Lookup` is used only in _class initializer_ - e.g. it may be needed at _build time_ of NI
- hopefully this NetBeans API isn't needed during runtime then

What is the failure you see when the dependency is missing?

JaroslavTulach replied:
If we only need `@ServiceProvider` annotation, then the `org-openide-util-lookup` library should only be `% "provided"`.

hubertp replied:
The dependency has already been part of `libraryDependencies` and marked with `provided` classifier, yes.

I was getting class not found exception.

hubertp replied:
Also, it's compile-time scope of `moduleDependecies` so that should be alright.

JaroslavTulach replied:
CCing @Akirathan. I am not really sure about "compile time only" behavior of `Compile / moduleDependencies`. As far as I know `Compile / moduleDependencies` puts the JAR on `--module-path` as _unnamed module_ (because `org-openide-util-lookup` isn't Java platform module). I don't think there is a reason for that - having it on classpath should be enough if we _only need its annotation processor_. Moreover having it on classpath with `provided` status prevents it to be included in the runtime JARs then.

It is a minor issue, but given:
- all the changes with `ServiceProvider` are only a workaround for https://github.com/oracle/graal/pull/10202
- maybe we could just create `META-INF/services/org.enso.languageserver.boot.LanguageServerRunner` manually
- such a file would be easier to delete once we update to new version of GraalVM

hubertp replied:
We do,
```
error: package org.openide.util.lookup is not visible
[error] @org.openide.util.lookup.ServiceProvider(service = LanguageServerApi.class)
[error]                  ^  (package org.openide.util.lookup is declared in the unnamed module, but module org.enso.language.server does not read it)
[error] 1 error
[error]                  ^
```

JaroslavTulach replied:
OK, thanks for trying. Let it be then.

JaroslavTulach commented on build.sbt:
```diff
@@ -2280,16 +2283,21 @@ lazy val `language-server` = (project in file("engine/language-server"))
       "org.eclipse.jgit"    % "org.eclipse.jgit"        % jgitVersion,
       "org.apache.tika"     % "tika-core"               % tikaVersion               % Test
     ),
+    libraryDependencies ++= GraalVM.modules.map(
```
Isn't `language-server` also a "client of GraalVM" - e.g. it shouldn't really need dependency on Truffle APIs? Not even runtime one?

JaroslavTulach commented on build.sbt:
```diff
@@ -3705,6 +3717,7 @@ lazy val `engine-runner` = project
               "-H:IncludeResources=.*Main.enso$",
               "-H:+AddAllCharsets",
               "-H:+IncludeAllLocales",
+              "-H:+UnlockExperimentalVMOptions",
```
What are the options we unlocking?

hubertp replied:
I wanted to get rid of all those warnings that native image builder spits out, thus hiding potentially useful information.
For example
```
[info] Warning: The option '-H:IncludeResources=.*Main.enso$' is experimental and must be enabled via '-H:+UnlockExperimentalVMOptions' in the future.
[info] Warning: Please re-evaluate whether any experimental option is required, and either remove or unlock it. The build output lists all active experimental options, including where they come from and possible alternatives. If you think an experimental option should be considered as stable, please file an issue
```

JaroslavTulach replied:
I see. OK, no problem with that. This purpose of the experimental option wasn't obvious from the static code change review.

JaroslavTulach commented on engine/common/src/main/java/org/enso/common/ContextFactory.java:
```diff
@@ -228,6 +232,19 @@ public Context build() {
 
+    if (engineOptions != null) {
+      // In AOT mode one must not use a shared engine; the latter causes issues when initializing
+      // message transport - it is set to `null`.
```
Interesting. I don't see a reason wny transport shouldn't work. Maybe it is a bug. Do we have a (small) reproducer that we could report to GraalVM guys and find out what they think?

JaroslavTulach commented on engine/language-server/src/main/java/module-info.java:
```diff
@@ -7,7 +7,7 @@
   requires flatbuffers.java;
   requires org.apache.commons.io;
-  requires org.graalvm.polyglot;
+  requires org.graalvm.truffle;
```
I don't think `language-server` should  be using Truffle API. I was wondering why we need that...

...Is it all because of `TruffleOptions.AOT`?  In such case we should just use the [boolean property](https://github.com/oracle/graal/blob/261c38a7a47dd87f13ec3cbe0fbb85cfdff17963/truffle/src/com.oracle.truffle.api/src/com/oracle/truffle/api/TruffleOptions.java#L123) and avoid Truffle API dependency.

JaroslavTulach commented on engine/language-server/src/main/java/module-info.java:
```diff
@@ -37,8 +37,8 @@
   requires org.enso.text.buffer;
   requires org.enso.task.progress.notifications;
   requires org.enso.ydoc.polyfill;
+  requires org.openide.util.lookup.RELEASE180;
```
I don't think this is needed. Some IDEs (surprisingly not NetBeans) like to add such dependency, but it was (almost) always incorrect change.


Now I see... the `@ServiceProvider` is used in `LanguageServerRunner`, but that should be just a _compile time depeendency_.

hubertp replied:
I was getting class not found exception otherwise.

JaroslavTulach replied:
That is surprising. What was the stack trace?

JaroslavTulach commented on engine/language-server/src/main/java/org/enso/languageserver/boot/LanguageServerRunner.java:
```diff
@@ -8,6 +8,7 @@
 import org.slf4j.event.Level;
 import scala.concurrent.ExecutionContext;
 
+@org.openide.util.lookup.ServiceProvider(service = LanguageServerApi.class)
```
This annotation is processed during _`javac` compile time_. It should not require changes to the runtime class path.

Btw. preferred way to register this class should be `module-info.java`, but yes, there is https://github.com/enso-org/enso/issues/7082#issuecomment-2428832039 which

> The fix has been merged (https://github.com/oracle/graal/pull/10202) and will be part of GraalVM 24.2.
>
>  We will also backport to GraalVM 23.1.

I assume this is a workaround for the above mentioned bug.

hubertp replied:
Interesting, I didn't even think of having it in `module-info.java`.

JaroslavTulach replied:
- It is in module-info since @Akirathan put it in [language-server module-info](https://github.com/enso-org/enso/blame/dee53b72d8b63cca174f6768af2b92a08a977fae/engine/language-server/src/main/java/module-info.java#L46)
- while working on #10823
- it just doesn't work in _native image_ mode because of https://github.com/oracle/graal/pull/10202
- thus all the changes with `ServiceLoader` are only a workaround

hubertp replied:
Ah, now I get it. Yes, it wasn't working in NI because of that bug.

JaroslavTulach commented on engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala:
```diff
@@ -317,7 +318,17 @@ class MainModule(serverConfig: LanguageServerConfig, logLevel: Level) {
   )
 
-  val builder = ContextFactory
+  var extraEngineOptions: util.HashMap[String, String] = null
```
This logic might also be encapsulated inside of `ContextFactory` itself. From outside of `ContextFactory`:
- we only specify whether to use `messageTransport` or not
- inside the factory does everything necessary to make `messageTransport` working
- which right now means creation of a dedicated engine in `AOT` mode.
- later, when the problem is investigated, enabling `messageTransport` might not need a such a trick anymore

hubertp replied:
I was thinking of doing it that way but decided against, so as to give more control over engine options. I don't have strong feelings about it so I can change it.

hubertp replied:
Hmm, it's also problematic because we make a dependency on `RuntimeServerInfo`.

JaroslavTulach replied:
I see. No huge problem with the current version, but 6e8c05b9b0d63f7cbba9393efea8b276f46cdd15 is even better!

JaroslavTulach commented on engine/language-server/src/main/scala/org/enso/languageserver/boot/MainModule.scala:
```diff
@@ -317,7 +318,17 @@ class MainModule(serverConfig: LanguageServerConfig, logLevel: Level) {
 
-  val builder = ContextFactory
+  var extraEngineOptions: util.HashMap[String, String] = null
+  if (TruffleOptions.AOT) {
```
Rather than depending on Truffle API, let's use

https://github.com/oracle/graal/blob/261c38a7a47dd87f13ec3cbe0fbb85cfdff17963/truffle/src/com.oracle.truffle.api/src/com/oracle/truffle/api/TruffleOptions.java#L123

to read the boolean value.

JaroslavTulach commented on engine/runner/src/main/java/org/enso/runner/Main.java:
```diff
@@ -215,6 +216,11 @@ private static Options buildOptions() {
+        cliOptionBuilder()
+            .longOpt(LANGUAGE_SERVER_NATIVE_OPTION)
+            .desc("Runs Language Server in native-image mode")
+            .build();
```
This looks like a serious end user facing API attempt...

Isn't it a bit too strong commitment to expose such a development only feature as a CLI option?

I am reading on...

JaroslavTulach commented on engine/runner/src/main/resources/META-INF/native-image/org/enso/runner/reflect-config.json:
```diff
@@ -939,6 +948,188 @@
+  "name":"org.slf4j.Logger",
+  "methods": [{"name":"getLogger", "parameterTypes":["java.lang.Class"]}]
+},
+{"name": "ch.qos.logback.core.status.InfoStatus"},
```
Couldn't we put these language server related resources to `language-server/src/main/resources/META-INF/native-image/...` directory?
- We don't want these definitions to have an effect if (ever) a CLI without a language server is built.
- Logically they belong to `language-server`

Each library can provide its own resources. They are composable. Keeping those that are needed for execution without `LanguageServerApi` implementation being registered/used here and those others needed for `LanguageServerRunner` elsewhere sounds architecturally tempting to me.

hubertp replied:
:+1:

JaroslavTulach commented on engine/runtime-instrument-runtime-server/src/main/java/org/enso/interpreter/instrument/runtime/server/RuntimeServerInstrument.java:
```diff
@@ -120,12 +120,13 @@ protected void onCreate(Env env) {
       MessageEndpoint client =
           env.startServer(URI.create(RuntimeServerInfo.URI), this.handler.endpoint());
       if (client != null) {
+        env.registerService(this);
```
Was there a race condition? Or is this a change in the logic where we only register this service when the `MessageEndpoint` is created?

Will not the infrastructure throw error when a service is defined by `@InstrumentRegistraiton` but not registered via `registerService` then?

hubertp replied:
I was getting NPEs when message transport was missing which wasn't particularly helpful and took a while to understand why NPE happens Now we don't attempt to register with `null`.

JaroslavTulach commented on engine/runtime-instrument-runtime-server/src/main/java/org/enso/interpreter/instrument/runtime/server/RuntimeServerInstrument.java:
```diff
@@ -108,7 +108,7 @@ public void onContextClosed(TruffleContext context) {}
     this.env = env;
-    env.registerService(this);
+
     if (TruffleOptions.AOT) {
```
The usage of `TruffleOptions.AOT` is correct here as this is a Truffle instrument and it needs dependency on Truffle API.

JaroslavTulach commented on lib/scala/project-manager/src/main/scala/org/enso/projectmanager/boot/Cli.scala:
```diff
@@ -45,6 +46,13 @@ object Cli {
+    val native: cli.Option = cli.Option.builder
+      .longOpt(NATIVE_OPTION)
+      .desc(
+        "(experimental) Attempts to use the native-image of any subprocess."
```
Can't we just use the `bin/enso` or `bin/enso.exe` if it is generated without any (even experimental) option? 

Btw. there already is support for `--jvm` option. We should align with it. Maybe just expose `--jvm` as an option to `project-manager` and just pass it on to `bin/enso`? 

Anyway the same can already be done with `ENSO_OPTS` environment variable and it may be enough for development purposes.

JaroslavTulach commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/runner/NativeJavaCommand.scala:
```diff
@@ -0,0 +1,20 @@
+    val dm       = new DistributionManager(env)
+    val execName = OS.executableName("enso")
+    val fullExecPath =
+      dm.paths.engines.resolve(version).resolve("bin").resolve(execName)
```
I'd remove all the support of `--native-image` options and just checked for existence of this `fullExecPath` binary (probably check it is not a bash shell) and if present just invoke it. It should be good enough for development purposes, but please note that it is also correct for the future:
- the [discussion entry](https://github.com/orgs/enso-org/discussions/10121#discussioncomment-10272101) outlines we should _replace launchers with binary ones_
- when `--jvm` option is specified the `bin/enso` or `bin\enso.exe` should switch to JVM mode themselves - already implemented since #11500 

E.g. preferring the `enso` binary if it exists is the right solution for the future.

JaroslavTulach commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/runner/Runner.scala:
```diff
@@ -188,23 +189,33 @@ class Runner(
+            "--module-path",
+            componentPath.toString,
+            "-m",
+            "org.enso.runner/org.enso.runner.Main"
```
Once we rely on `bin/enso` native binaries, we will no longer be able to access `Main` directly. Such a _"JVM switch"_ will be (and already is) handled by `bin/enso` when `--jvm` option is passed to them. E.g. this code should disappear at the end, and project manager shall always call into `bin/enso`.

JaroslavTulach reviewed (approved):
- I am glad to see language server startup improvement
- that's what _native image_ technology is good for and I am glad to see it delivers on its promise
- for a while I was confused by the dependency changes
- looks like it is mostly to access `TruffleOptions.AOT` from the non-Truffle part of our codebase
- please keep Truffle API out of the dependencies of _launcher, boot, etc._ projects
- use [Boolean.getBoolean("com.oracle.graalvm.isaot")](https://github.com/oracle/graal/blob/master/truffle/src/com.oracle.truffle.api/src/com/oracle/truffle/api/TruffleOptions.java#L123) directly
- I suggest to avoid introduction of `--native-...` CLI options
    - we may not need them at all at this stage
    - just prefer `bin/enso` or `bin\enso.exe` native binary when it exists
    - there already is well functioning `--jvm` option
    - we may want to expose it from project manager CLI, but
    - it is already accessible via `ENSO_OPTS` and that may be enough for now
- I suggest to separate separate `*-config.json` and put them into `language-server` source tree
- I am glad to see the #10121 work moving forward

hubertp committed:
Drop TruffleOptions dependency

Despite being advertised as such, native image doesn't set
`com.oracle.graalvm.isaot` and it has to be provided separately if we
don't want to use `TruffleOptions`.

hubertp committed:
Move NI config to LS resources

JaroslavTulach commented on engine/language-server/src/main/resources/META-INF/native-image/org/enso/languageserver/serialization-config.json:
```diff
@@ -0,0 +1,59 @@
+{
```
+1

hubertp committed:
Remove automatically added module dependency

hubertp committed:
Move AOT Context-creation logic to factory

Context creation for AOT needs to define a separate Engine configuration
rather than using a shared engine to be able to use message transport.

hubertp committed:
PR review

JaroslavTulach commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/runner/NativeExecCommand.scala:
```diff
@@ -0,0 +1,32 @@
+    engine: Engine,
+    jvmSettings: JVMSettings
+  ): Seq[String] =
+    Seq("-Dcom.oracle.graalvm.isaot=true")
```
Interesting. Why do we need to provide this argument?
- I would expect it is only useful at image built time
- or that it is provided by default
- btw. [Native Image API offers some checks](https://www.graalvm.org/sdk/javadoc/org/graalvm/nativeimage/ImageInfo.html) like this too

E.g. if we can access `org.graalvm.nativeimage.ImageInfo` then we can query _"properly"_ and even distinguish between build time and runtime time.

hubertp replied:
The property is completely missing during runtime, hence not allowing me to provide different Context configuration.

JaroslavTulach replied:
> access `org.graalvm.nativeimage.ImageInfo` ... distinguish between build time and runtime time.

Another incarnation of this problem [found here](https://github.com/enso-org/enso/pull/12443#discussion_r1985989978)

JaroslavTulach commented:
The description should be updated as

> 
>    engine-runner/buildNativeImage
>    Run PM with --native-language-server
>

may no longer reflect the reality after 3ce2577

hubertp committed:
Merge branch 'develop' into wip/hubert/11721-language-server-native

hubertp committed:
Merge branch 'develop' into wip/hubert/11721-language-server-native

JaroslavTulach commented:
Congratulation on merging and making another step towards using _native binaries_ as GUI backend. Notes:
- generating `enso` now takes  5m 48s on my Ubuntu. Used to be 3m.
- the `enso` executable size is now 350MB. Used to be ~200MB.

This is the **top ten** of included packages:
```
Top 10 origins of code area:
  21.50MB js-language-24.0.0.jar
  15.69MB java.base
   6.75MB org.graalvm.truffle
   5.48MB jdk.internal.vm.compiler
   4.81MB java.xml
   4.15MB akka-http-core_2.13-10.2.10.jar
   4.14MB poi-5.2.3.jar
   3.87MB icu4j-24.0.0.jar
   3.34MB org.enso.languageserver.protocol.json
```
- it is surprising to see _JavaScript back_! 
- it was removed in #11279
- according to #10121 we don't want/need Python neither JavaScript in the image yet
- removing Graal.js would speed build times
- removing Graal.js would lower the size of the `enso` executable significantly
- reported as #12018

JaroslavTulach commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/runner/JavaExecCommand.scala:
```diff
@@ -0,0 +1,69 @@
+        "--module-path",
+        componentPath.toString,
+        "-m",
+        "org.enso.runner/org.enso.runner.Main"
```
Rather referencing `org.enso.runner.Main` class, the `project-manager` should just use `NativeExecCommand` and pass in `--jvm` - once #12302 is fixed.
